.ado/publish.yml

```diff
@@ -1,45 +1,195 @@
+        sudo apt-get update
+        sudo apt-get install -y --no-install-recommends \
+          qemu qemu-system-misc qemu-user-static qemu-user
+      fi
+
+      chmod +x pip/build_manylinux.sh
+
+      docker run \
+        --env VER_MAJOR=$(Ver.Major) --env VER_MINOR=$(Ver.Minor) --env VER_PATCH=$(Ver.Patch) --env ARCH=$ARCH \
+        --rm -i -v `pwd`:/io \
+        quay.io/pypa/manylinux_2_28_$ARCH \
+        /io/pip/build_manylinux.sh
+
+      ls target/wheels
+    displayName: Build Platform-Dependent Py Packages, Linux
+    condition: eq(variables['Agent.OS'], 'Linux')
+
+  - script: |
+      python build.py --jupyterlab
+    displayName: Build Platform-Independent Py Packages
+    condition: eq(variables['Agent.OS'], '${{ variables.PlatfIndepPyPackBuildPlatf }}')
 
   - script: |
-      vsce publish --pre-release --pat $(PAT) 2>&1 > pub.log
+      dir target\wheels\*
+    displayName: List Py Packages on Win
+    condition: eq(variables['Agent.OS'], 'Windows_NT')
+
+  - script: |
+      ls target/wheels/*
+    displayName: List Py Packages on non-Win
+    condition: ne(variables['Agent.OS'], 'Windows_NT')
+
+  - publish: $(System.DefaultWorkingDirectory)/target/wheels
+    artifact: Wheels.Mac
+    displayName: Upload Python Artifacts Mac
+    condition: eq(variables['Agent.OS'], 'Darwin')
+
+  - publish: $(System.DefaultWorkingDirectory)/target/wheels
+    artifact: Wheels.Win
+    displayName: Upload Python Artifacts Win
+    condition: eq(variables['Agent.OS'], 'Windows_NT')
+
+  - publish: $(System.DefaultWorkingDirectory)/target/wheels
+    artifact: Wheels.Linux.${{ variables['arch'] }}
+    displayName: Upload Python Artifacts Linux
+    condition: eq(variables['Agent.OS'], 'Linux')
+
+
+- job: "Publish"
+  dependsOn: Build
+  pool:
+    vmImage: 'windows-latest'
+
+  steps:
+  - download: current
+    artifact: Wheels.Win
+    displayName: Download Python Artifacts Win
+
+  - download: current
+    artifact: Wheels.Mac
+    displayName: Download Python Artifacts Mac
+
+  - download: current
+    artifact: Wheels.Linux.x86_64
+    displayName: Download Python Artifacts Linux
+
+  - download: current
+    artifact: Wheels.Linux.aarch64
+    displayName: Download Python Artifacts Linux
+
+  - script: |
+      mkdir                               target\wheels
+      move ..\Wheels.Linux.x86_64\*.whl   target\wheels
+      move ..\Wheels.Linux.aarch64\*.whl  target\wheels
+      move ..\Wheels.Win\*.whl            target\wheels
+      move ..\Wheels.Mac\*.whl            target\wheels
+      dir                                 target\wheels\*
+    displayName: Move Py Artifacts to Publishing Dir
+
+  - download: current
+    artifact: VSIX
+    displayName: Download VSCode Extension Artifact
+
+  - script: |
+      npm install -g @vscode/vsce
+    displayName: Install Prereqs for VSCode Ext Publishing
+
+
+  # NB: Publishing must be the last thing (done only if all previous steps are successful).
+
+  - task: EsrpRelease@2
```

Comment:
Just had another thought... I wonder if we should have two separate pipelines, one for packaging (i.e. building the wheels), and the other for actually publishing. The reason is that we may want to download the built wheels to do some local validation before we actually publish to PyPi.org.

Perhaps the first pipeline (packaging) could just produce .whl files as artifacts, and the second pipeline could grab those artifacts and publish them?

Reply:
It is possible, but eventually I believe the validation will be automated too. In any case the first thing that I would consider is breaking-up the pipeline into stages - building, validating, publishing, - and, if needed, when kicking-off the pipeline, to specify, which stages to run (e.g., all except publishing). After the building and validation stages have run successfully, we can "re-run" the publishing stage.
The other option could be in the same way as with qdk.release pipeline - the publishing stage requires manual approval (button press). If approval does not happen within time-out, then the publishing stage is not run. But later on (after manual validation) it is "re-run" and approved, ending up in publishing the built and tested artifacts. At one point I tried to figure out how exactly the qdk.release pipeline implements that approval, and at a glance I didn't see that approval in an explicit form. Probably that approval is auto-added hiddenly since the stage uses the nuget.org publishing task (and certain authentication mechanisms take place under the hood). If needed, I'll find out for sure.

Once we get closer to actual publishing, we'll see more clearly what we need. Implementing that seems to me a relatively straightforward task.